vsock-bridge/src/socket.cpp

```diff
@@ -208,27 +210,49 @@ namespace vsockio
 		{
 			_inputClosed = true;
 			_outputClosed = true;
-			Logger::instance->Log(Logger::DEBUG, "socket shutdown, fd=", _fd);
 
-			_impl->close(_fd);
+			if (_poller)
+			{
+				// epoll is meant to automatically deregister sockets on close, but apparently some systems
+				// have bugs around this, so do it explicitly
+				Logger::instance->Log(Logger::DEBUG, "[socket] remove from poller (fd=", _fd, ")");
+				_poller->remove(_fd);
+			}
+
+			Logger::instance->Log(Logger::DEBUG, "[socket] close, fd=", _fd);
+			_impl.close(_fd);
 			if (_peer != nullptr)
 			{
-				_peer->onPeerShutdown();
+				_peer->onPeerClosed();
 			}
 		}
 	}
 
-	void Socket::onPeerShutdown()
+	void Socket::onPeerClosed()
 	{
 		if (!closed())
 		{
-			Logger::instance->Log(Logger::DEBUG, "[socket] sending termination from (fd=", _fd, ")");
+			Logger::instance->Log(Logger::DEBUG, "[socket] sending termination for (fd=", _fd, ")");
 			std::unique_ptr<Buffer> termination{ BufferManager::getEmptyBuffer() };
-			queue(termination);
-			bool _;
-			writeToOutput(_);
+			queue(std::move(termination));
+
+			// force process the queue
+			_outputReady = true;
+			writeToOutput();
 		}
 	}
 
-	Socket::~Socket() {}
+	Socket::~Socket()
+	{
+		if (!closed())
+		{
+			Logger::instance->Log(Logger::WARNING, "[socket] closing on destruction (fd=", _fd, ")");
+			close();
+		}
+
+		if (_peer != nullptr)
+		{
+			_peer->setPeer(nullptr);
```

Comment:
Shall we prevent `Peer::onInputReady` and `Peer::onOutputReady` from panicking from assertion failure if we use `nullptr` to indicate a closed peer?

Reply:
Both sockets should be destroyed together when the parent channel is destroyed. Those callbacks should not be invoked while this is taking place.